internal/dag/gatewayapi_processor.go

```diff
@@ -195,39 +196,111 @@ func isSecretRef(certificateRef *gatewayapi_v1alpha1.LocalObjectReference) bool
+
 	var hosts []string
 	var errors []error
-	if len(route.Spec.Hostnames) == 0 {
+
+	// Determine the hosts on the route, if no hosts
+	// are defined, then set to "*". If the listenerHostname is defined,
+	// then the route must match the Gateway hostname.
+	if len(route.Spec.Hostnames) == 0 && listenerHostname == nil {
 		hosts = append(hosts, "*")
 		return hosts, nil
 	}
 
+	if listenerHostname != nil {
+		if string(*listenerHostname) != "*" {
+
+			// Validate listener hostname.
+			if err := validHostName(string(*listenerHostname)); err != nil {
+				return hosts, []error{err}
+			}
+
+			if len(route.Spec.Hostnames) == 0 {
+				hosts = append(hosts, string(*listenerHostname))
+				return hosts, nil
+			}
+		}
+	}
+
 	for _, host := range route.Spec.Hostnames {
 
 		hostname := string(host)
-		if isIP := net.ParseIP(hostname) != nil; isIP {
-			errors = append(errors, fmt.Errorf("hostname %q must be a DNS name, not an IP address", hostname))
+
+		// A "*" hostname matches anything.
+		if hostname == "*" {
+			hosts = append(hosts, hostname)
+			continue
+		}
+
+		// Validate the hostname.
+		if err := validHostName(hostname); err != nil {
+			errors = append(errors, err)
 			continue
 		}
-		if strings.Contains(hostname, "*") {
-			if errs := validation.IsWildcardDNS1123Subdomain(hostname); errs != nil {
-				errors = append(errors, fmt.Errorf("invalid hostname %q: %v", hostname, errs))
+
+		if listenerHostname != nil {
+			const wildcardString string = "[a-z0-9]([-a-z0-9]*[a-z0-9])*."
+			lhn := string(*listenerHostname)
+
+			// A "*" hostname matches anything.
+			if lhn == "*" {
+				hosts = append(hosts, hostname)
 				continue
 			}
-		} else {
-			if errs := validation.IsDNS1123Subdomain(hostname); errs != nil {
-				errors = append(errors, fmt.Errorf("invalid listener hostname %q: %v", hostname, errs))
-				continue
+
+			if strings.Contains(lhn, "*") {
+
+				// Build a regex to match the hostname against the wildcard listener hostname.
+				wildcardDNS1123SubdomainRegexp := regexp.MustCompile("^" + strings.Replace(lhn, "*", wildcardString, 1) + "$")
+
+				// Validate the wildcard gateway listener hostname matches the route hostname.
+				if !wildcardDNS1123SubdomainRegexp.MatchString(hostname) {
+					errors = append(errors, fmt.Errorf("gateway hostname %q does not match route hostname %q", lhn, hostname))
+					continue
+				}
+			} else if strings.Contains(hostname, "*") {
+
+				// Build a regex to match the hostname against the wildcard route hostname.
+				wildcardDNS1123SubdomainRegexp := regexp.MustCompile("^" + strings.Replace(hostname, "*", wildcardString, 1) + "$")
+
+				// Validate the wildcard route hostname matches gateway listener hostname.
+				if !wildcardDNS1123SubdomainRegexp.MatchString(lhn) {
+					errors = append(errors, fmt.Errorf("route hostname %q does not match gateway hostname %q", hostname, lhn))
+					continue
+				}
```

Comment:
can both the listener and route hostname be a wildcard, e.g. `*.foo.com` for both? If so then before doing the wildcard matching we should check for string equality between the listener and route host and add it if so.

Reply:
Yup nice catch, I fixed that up and added a test for it.